lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -1668,58 +1683,121 @@ struct U3ToRotations : public OpRewritePattern<quake::U3Op> {
   }
 };
 
-} // namespace
-
 //===----------------------------------------------------------------------===//
-// Populating pattern sets
+// Metadata for all decomposition patterns
+//
+// IMPORTANT: Make sure to add new patterns to the end of the `patternMetadata`
+// map below.
 //===----------------------------------------------------------------------===//
 
-void cudaq::populateWithAllDecompositionPatterns(RewritePatternSet &patterns) {
-  // clang-format off
-  patterns.insert<
-    // HOp patterns
-    HToPhasedRx,
-    CHToCX,
-    // SOp patterns
-    SToPhasedRx,
-    SToR1,
-    // TOp patterns
-    TToPhasedRx,
-    TToR1,
-    // XOp patterns
-    CXToCZ,
-    CCXToCCZ,
-    XToPhasedRx,
-    // YOp patterns
-    YToPhasedRx,
-    CYToCX,
-    // ZOp patterns
-    CZToCX,
-    CCZToCX,
-    ZToPhasedRx,
-    // R1Op patterns
-    CR1ToCX,
-    R1ToPhasedRx,
-    R1ToRz,
-    R1ToU3,
-    R1AdjToR1,
-    // RxOp patterns
-    CRxToCX,
-    RxToPhasedRx,
-    RxAdjToRx,
-    // RyOp patterns
-    CRyToCX,
-    RyToPhasedRx,
-    RyAdjToRy,
-    // RzOp patterns
-    CRzToCX,
-    RzToPhasedRx,
-    RzAdjToRz,
-    // Swap
-    SwapToCX,
-    // U3Op
-    U3ToRotations,
-    ExpPauliDecomposition
-  >(patterns.getContext());
-  // clang-format on
+struct PatternMetadata {
+  llvm::StringRef sourceGate;
+  llvm::SmallVector<llvm::StringRef> targetGates;
+  std::function<std::unique_ptr<mlir::RewritePattern>(mlir::MLIRContext *)>
+      factoryFn;
+};
+
+template <typename Pattern>
+PatternMetadata makePatternMetadata(StringRef sourceGate,
+                                    SmallVector<StringRef> targetGates) {
+  return PatternMetadata{
+      .sourceGate = sourceGate,
+      .targetGates = targetGates,
+      .factoryFn =
+          [](MLIRContext *ctx) { return RewritePattern::create<Pattern>(ctx); },
+  };
+}
+const std::map<StringRef, PatternMetadata> &getAllPatternMetadata() {
+  /// Each entry documents:
+  /// - the debug name of the pattern (must match getDebugName() of the
+  /// pattern),
+  /// - the source gate, i.e. the gate that the pattern matches and decomposes,
+  /// - the target gates, i.e. the gates that the pattern creates. This must
+  ///   include all gates that may be created by any instance of the pattern.
+  static const std::map<StringRef, PatternMetadata> patternMetadata{
+      // TODO: "SToR1", "TToR1", "R1ToU3", "U3ToRotations" can be generalised to
```

Comment:
The pattern rewrites enumerated in the TODO also match on the `Op`s when they have an unbounded number of controls. However, reasoning about this is hard: we would need to introduce functional relationships between `n` the number of controls in the matched op and `m`, the number of controls on (one of) the ops in the output ops.

I've simplified this here by only advertising `n=0` in these metadata properties.